src/Microsoft.Kusto.ServiceLayer/DataSource/KustoDataSource.cs

```diff
@@ -626,93 +612,311 @@ internal IEnumerable<ColumnInfo> GetColumnMetadata(string databaseName)
+            AddToFolderMetadata(functionFolderKey, new List<FolderMetadata> {functionFolder});
+
+            // create each folder to hold functions
+            var functionsGroupByFolder = functionInfos
+                .GroupBy(x => x.Folder, StringComparer.OrdinalIgnoreCase)
+                .ToList();
+
+            var folders = new Dictionary<string, Dictionary<string, FolderMetadata>>();
+            
+            foreach (var functionGroup in functionsGroupByFolder)
+            {
+                // skip functions with no folder
+                if (string.IsNullOrWhiteSpace(functionGroup.Key))
+                {
+                    continue;
+                }
+                
+                var subFolders = functionGroup.Key.Replace(@"\", @"/").Split(@"/");
+                var topFolder = subFolders.First();
+
+                var folderKey = functionFolder.Urn;
+                var folder = DataSourceFactory.CreateFolderMetadata(objectMetadata, folderKey, topFolder);
+                folders.SafeAdd(folderKey, folder);
                 
-                if (!columns.Any())
+                for (int i = 1; i < subFolders.Length; i++)
                 {
-                    tableMetadata[databaseName] = Enumerable.Empty<DataSourceObjectMetadata>();
+                    folderKey = $"{folderKey}.{subFolders[i - 1]}";
+                    var subFolder = DataSourceFactory.CreateFolderMetadata(objectMetadata, folderKey, subFolders[i]);
+                    folders.SafeAdd(folderKey, subFolder);
                 }
+            }
 
-                var tableGroups = columns
-                    .GroupBy(row => row.Table, StringComparer.Ordinal); // case-sensitive
-
-                // Get all table names
-                tableMetadata[databaseName] = tableGroups
-                    .Select(grouping => new TableMetadata
-                            {
-                                ClusterName = this.ClusterName,
-                                DatabaseName = databaseName,
-                                MetadataType = DataSourceMetadataType.Table,
-                                MetadataTypeName = DataSourceMetadataType.Table.ToString(),
-                                Name = grouping.Key, // table name
-                                PrettyName = grouping.Key,
-                                Urn = $"{this.ClusterName}.{databaseName}.{grouping.Key}"
-                            })
-                    .OrderBy(row => row.Name, StringComparer.Ordinal); // case-sensitive
+            foreach (var folder in folders)
+            {
+                AddToFolderMetadata(folder.Key, folder.Value.Values.ToList());
+            }
+            
+            SetFunctionMetadata(objectMetadata.Name, functionFolder.Name, functionsGroupByFolder);
+        }
+
+        private void SetColumnMetadata(string databaseName, IEnumerable<IGrouping<string, ColumnInfo>> columnsGroupByTable)
```

Comment:
It would be good to add a comment header explaining the purpose of the function what each param is for each function.